mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -4744,6 +4933,364 @@ def _get_orderby_clauses(order_by_list, session):
+                .subquery()
+            )
+            id_source = subquery.c.run_uuid
+        elif target_type == EntityAssociationType.TRACE:
+            subquery = (
+                session.query(SqlTraceInfo.request_id)
+                .join(SqlExperiment, SqlTraceInfo.experiment_id == SqlExperiment.experiment_id)
+                .filter(SqlExperiment.workspace == workspace)
+                .subquery()
+            )
+            id_source = subquery.c.request_id
+        elif target_type == EntityAssociationType.EVALUATION_DATASET:
+            subquery = (
+                session.query(SqlEvaluationDataset.dataset_id)
+                .filter(SqlEvaluationDataset.workspace == workspace)
+                .subquery()
+            )
+            id_source = subquery.c.dataset_id
+        else:
+            return query
+
+        return query.filter(id_column.in_(select(id_source)))
+
+    def _add_fields(self, instance):
+        instance = super()._add_fields(instance)
+        if hasattr(instance, "workspace"):
+            instance.workspace = self._get_active_workspace()
+        return instance
+
+    def _get_workspace_provider_instance(self):
+        if self._workspace_provider is None:
+            workspace_uri = workspace_utils.resolve_workspace_store_uri(tracking_uri=self.db_uri)
+            self._workspace_provider = get_workspace_store(workspace_uri=workspace_uri)
+        return self._workspace_provider
+
+    @staticmethod
+    def _artifact_path_segments(uri: str | None) -> list[str]:
+        if not uri:
+            return []
+        parsed = urlparse(uri)
+        path = parsed.path if parsed.scheme else uri
+        return [segment for segment in path.split("/") if segment]
+
+    def _validate_artifact_isolation_constraints(self) -> None:
+        """Ensure the default artifact root and existing artifacts do not occupy reserved paths."""
+
+        segments = self._artifact_path_segments(self.artifact_root_uri.rstrip("/"))
+        if segments and segments[-1] == "workspaces":
+            raise MlflowException(
+                "Cannot enable workspace mode because the default artifact root "
+                + f"{self.artifact_root_uri} ends with the reserved 'workspaces' segment. "
+                + "Choose a different artifact root before enabling workspaces.",
+                error_code=INVALID_STATE,
+            )
+        if len(segments) >= 2 and segments[-2] == "workspaces":
+            raise MlflowException(
+                "Cannot enable workspace mode because the default artifact root "
+                + f"{self.artifact_root_uri} is already scoped under the reserved "
+                + "'workspaces/<name>' prefix. Configure a different artifact root before enabling "
+                + "workspaces.",
+                error_code=INVALID_STATE,
+            )
+
+        reserved_prefix = append_to_uri_path(self.artifact_root_uri, "workspaces").rstrip("/") + "/"
```

Comment:
We shouldn't be string munging unix paths. This definitely won't work on NTFS. Use pathlib.